# Assignment for week 5

Use the following table to provide us with

|name | exam number|
|----|----|
|Jesper Krauth| u766044|

In the lecture we considered a two period bargaining model. In this assignment we are going to extend this to 3 periods.

Pay attention to what you do here. The python part of the final assignment will be the extension to $n$ bargaining periods; but let us not worry about this yet.

We need the following library:

In [12]:
import numpy as np

As we are looking at more periods here, we need a finer grid of offers. We want all offers between 0.0 and 10.0 (including 10.0) with step 0.1 between offers.

**1)** Define the list `offers` using `np.arrange`:

In [13]:
min_offer = 0.0
max_offer = 10.0
step = 0.1
offers = np.arange(min_offer,max_offer+0.1,0.1)

**2)** Copy the function `accept_offer` from the lecture:

In [14]:
def accept_offer(offer,your_outside_option):
    accept = (offer >= your_outside_option)
    return accept

**3)** Copy the function `make_offer` from the lecture:

In [15]:
def make_offer(your_outside_option,other_outside_option):
    profits = [(max_offer-offer)*accept_offer(offer,other_outside_option) for offer in offers]
    max_profit = max(profits)
    max_index = profits.index(max_profit)
    if max_profit >= your_outside_option:
        your_offer = offers[max_index]
        your_profit = max_offer-offers[max_index]
        other_profit = offers[max_index]
    else:
        your_offer = -1 # no offer is made
        your_profit = your_outside_option
        other_profit = other_outside_option
    return your_offer, your_profit, other_profit

Pay attention to what `make_offer` returns and in which order. You need this below.

In this assignment, we are looking at the three period bargaining problem. We want to use the function `make_offer` to keep track of the outside options without defining a new dictionary `outside_options` (as we did in the lecture).

As in the lecture, we will use backward induction to solve this. That is, we first consider what happens in period 3 --in case the offers made in periods 1 and 2 are rejected. Once we know what will happen in period 3, we can move back to period 2 and then to period 1.

To do this, we create a dictionary `offer`. 

**4)** Explain what the following code does (e.g. what is `offer[4]`?).

In [16]:
delta = 0.9
offer = {}
offer[4] = [0,0,0]
offer[3] = make_offer(delta*offer[4][2],delta*offer[4][1])



Explanation: This code makes sure that only three rounds are played. For every 4th round (offer[4]), there will be no deal at all, regardless of the offer made. Hence, offer[4] = [0,0,0]. The delta code gives us the average change in the offer (which can be seen in the code for offer[3], when the new offer will be delta times the offer).

**5)** Using the logic above, define `offer[2]` and `offer[1]`:

In [17]:
offer[2] = make_offer (delta*offer[3][2],delta*offer[3][1])
offer[1] = make_offer (delta*offer[2][2],delta*offer[2][1])

**6)** What are the payoffs for players 1 and 2 in a three period version of the bargaining model in the lecture?

**7)** In which period will the players reach an agreement?

In [18]:
# Question 6
offer

{1: (0.90000000000000002, 9.0999999999999996, 0.90000000000000002),
 2: (9.0, 1.0, 9.0),
 3: (0.0, 10.0, 0.0),
 4: [0, 0, 0]}

**Question 6 answer:** The data above shows the 4 rounds that will be played. For each round, there are 3 values given. The first value is the offer made by one of the players, the second value is the profit of the player who made the offer, and the third value is the profit of the other player. So for example, in round 2, the offer is 9.0, the profit of the player who made the offer is 1.0 and the profit of the other player is 9.0.

**Question 7 answer:** The game will end in round 1. If we use the backward induction to solve this, we will see that in round 4 no one makes a profit. Hence, we will take a look at round 3. In this round, player 2 will receive a profit 0.0. But obviously, player 2 wants to make a profit, too, so we take a look at round 2. In this round, player 2 receives a profit of 1.0 and player 1 receives a profit of 9.0. But player 1 obviously wants his profit to be as high as possible. If we look at round 1, player 1 receives a profit of (approximately) 9.1, which is higher than the 9.0 profit of round 2. So, the game will end at round 1.